### PR TITLE
Scaffolding Implemententation 

### PR DESCRIPTION
**Pull request by pgiacomo69**
Implemented Scaffolding, Added Blob Mapping from Db to Model

### CODE REVIEW

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 1]
```diff
@@ -14,62 +14,446 @@
...
 
-			var connectionStartedOpen = _connection.State == ConnectionState.Open;
+			var connectionStartedOpen = connection.State == ConnectionState.Open;
 			if (!connectionStartedOpen)
 			{
-				_connection.Open();
+				connection.Open();
 			}
+
+#warning Finish
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:60]: Should be probably removed. I suppose it's finished. :D

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 2]
```diff
@@ -14,62 +14,446 @@
...
+			foreach (var table in tables)
+			{
+				using (var command = connection.CreateCommand())
+				{
+					command.CommandText = string.Format(GetColumnsQuery, table.Name);
+					using (var reader = command.ExecuteReader())
+					{
+						while (reader.Read())
+						{
+							String name = reader["COLUMN_NAME"].ToString();
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:212]: Use aliases, i.e. `String` => `string`, etc.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 3]
```diff
@@ -14,62 +14,446 @@
...
+    WHEN 37 THEN 'VARCHAR(' || (TRUNC(F.RDB$FIELD_LENGTH / CH.RDB$BYTES_PER_CHARACTER)) || ')'
+    WHEN 40 THEN 'CSTRING' || (TRUNC(F.RDB$FIELD_LENGTH / CH.RDB$BYTES_PER_CHARACTER)) || ')'
+    WHEN 45 THEN 'BLOB_ID'
+    WHEN 261 THEN 'BLOB SUB_TYPE ' || F.RDB$FIELD_SUB_TYPE 
+    ELSE 'RDB$FIELD_TYPE: ' || F.RDB$FIELD_TYPE || '?'
+    END as STORE_TYPE,
+    ch.RDB$CHARACTER_SET_NAME as CHARACTER_SET_NAME,
+    F.rdb$description as COLUMN_COMMENT,
+    COALESCE(RF.RDB$IDENTITY_TYPE, 0)   as auto_generated
+
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:186]: Extra lines (also check other queries).

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 4]
```diff
@@ -14,62 +14,446 @@
...
+    WHEN 35 THEN 'TIMESTAMP'
+    WHEN 37 THEN 'VARCHAR(' || (TRUNC(F.RDB$FIELD_LENGTH / CH.RDB$BYTES_PER_CHARACTER)) || ')'
+    WHEN 40 THEN 'CSTRING' || (TRUNC(F.RDB$FIELD_LENGTH / CH.RDB$BYTES_PER_CHARACTER)) || ')'
+    WHEN 45 THEN 'BLOB_ID'
+    WHEN 261 THEN 'BLOB SUB_TYPE ' || F.RDB$FIELD_SUB_TYPE 
+    ELSE 'RDB$FIELD_TYPE: ' || F.RDB$FIELD_TYPE || '?'
+    END as STORE_TYPE,
+    ch.RDB$CHARACTER_SET_NAME as CHARACTER_SET_NAME,
+    F.rdb$description as COLUMN_COMMENT,
+    COALESCE(RF.RDB$IDENTITY_TYPE, 0)   as auto_generated
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:185]: Inconsistent formatting after `as` (also check other queries).

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 5]
```diff
@@ -14,62 +14,446 @@
...
+					using (var reader = command.ExecuteReader())
+					{
+						while (reader.Read())
+						{
+							String name = reader["COLUMN_NAME"].ToString();
+							String defaultValue = reader["COLUMN_DEFAULT"].ToString();
+							Boolean nullable = !Convert.ToBoolean(reader["NOT_NULL"]);
+							Boolean autoGenerated = Convert.ToBoolean(reader["AUTO_GENERATED"]);
+							String storeType = reader["STORE_TYPE"].ToString();
+							// String dataType = reader["DATA_TYPE"].ToString();
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:217]: If it's not used, I prefer to remove it. Any reason to keep it?

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 6]
```diff
@@ -14,62 +14,446 @@
...
+					}
+				}
+			}
+		}
+
+		private const string GetConstraintsQuery = @"SELECT
+    trim(drs.rdb$constraint_name) as CONSTRAINT_NAME,
+    trim(drs.RDB$RELATION_NAME) as TABLE_NAME,
+    trim(mrc.rdb$relation_name) AS REFERENCED_TABLE_NAME,
+    /* LIST(distinct trim(dis.rdb$field_name)||'#'||dis.rdb$field_position,',') AS SRC_COLUMN_NAME,
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:386]: Remove.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 7]
```diff
@@ -14,62 +14,446 @@
...
+       LIST(distinct trim(mis.rdb$field_name)||'#'||mis.rdb$field_position,',') AS PRINC_COLUMN_NAME, */
+    ( select list(trim(di.rdb$field_name)||'|'||trim(mi.rdb$field_name))
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:388]: Inconsistent.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs [hunk 8]
```diff
@@ -211,7 +211,14 @@ protected override void Generate(CreateIndexOperation operation, IModel model, M
 		}
 
 		protected override void Generate(DropIndexOperation operation, IModel model, MigrationCommandListBuilder builder, bool terminate = true)
-			=> base.Generate(operation, model, builder, terminate);
+		{
```
[the remark this answers is not in the document]
Comment (pgiacomo69) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs:214]: "base.Generate(DropIndexOperation..." causes a System.NotImplementedException .
The problem happens if you drop or change an index and then try to migrate the database.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs [hunk 9]
```diff
@@ -14,62 +14,446 @@
...
+		private string GetDefaultSchema(DbConnection connection)
 		{
-			_databaseModel = new DatabaseModel();
-			_connection = null;
-			_serverVersion = null;
+			return null;
+		}
+
+		private static Func<string, string, bool> GenerateTableFilter(
+			IReadOnlyList<string> tables,
```
[the remark this answers is not in the document]
Comment (pgiacomo69) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Scaffolding/Internal/FbDatabaseModelFactory.cs:97]: Sorry for the inconvenient, and thank you for the patience in reading the code, I will fix it

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cincuranet) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs:214]: Ahh, I missed that. To keep things clean, I would prefer to have this change to have in separate PR.

**Pull request closed without merging**